tests/azure_functions/test_function_app.py

```diff
@@ -1,15 +1,378 @@
+import pytest
+import requests.exceptions
+
+import azure.functions as func
+from azure_functions.function_app import (
+    alert_to_slack,
+    build_slack_message,
+    fetch_search_results,
+    format_alert_date,
+    format_item,
+    format_raw_stack,
+    format_search_results,
+    get_details_string,
+    health_check,
+    select_search_results,
+    send_to_slack,
+    validate_function_key,
+)
+
+
+@pytest.mark.parametrize(
+    "key, value, expected_output",
+    [
+        ("Severity", "Sev1", "*Severity*: Sev1"),
+        ("", "Value", "**: Value"),
+        ("Key", None, "*Key*: N/A"),
+    ],
+)
+def test_format_item(key, value, expected_output):
+    result = format_item(key, value)
+    assert result == expected_output
+
+
+@pytest.mark.parametrize(
+    "data, expected_output",
+    [("2023-01-01T12:00:00.12345Z", "2023-01-01T12:00:00Z"), ("", "N/A"), (None, "N/A"), ("invalid-datetime", "N/A")],
+)
+def test_format_alert_date(data, expected_output):
+    result = format_alert_date(data)
+    assert result == expected_output
+
+
+def test_format_raw_stack_long():
+    stack = """
+        Traceback (most recent call last):
+          File "app.py", line 10, in <module>
+            my_func()
+          File "app.py", line 7, in my_func
+            raise ValueError("An error")
+        ValueError: An error
+        Traceback (most recent call last):
+          File "app.py", line 10, in <module>
+            my_func()
+          File "app.py", line 7, in my_func
+            raise ValueError("An error")
+        ValueError: An error
+        Traceback (most recent call last):
+          File "app.py", line 10, in <module>
+            my_func()
+          File "app.py", line 7, in my_func
+            raise ValueError("An error")
+        ValueError: An error
+        Traceback (most recent call last):
+          File "app.py", line 10, in <module>
+            my_func()
+          File "app.py", line 7, in my_func
+            raise ValueError("An error")
+        ValueError: An error
+        """
+    dedented_stack = textwrap.dedent(stack).strip()
+    lines = dedented_stack.splitlines()
+    expected_first_10 = "\n".join(lines[:10])
+    expected_last_10 = "\n".join(lines[-10:])
+    expected_result = f"{expected_first_10}\n ... \n{expected_last_10}"
+
+    result = format_raw_stack(stack)
+
+    assert result == expected_result
+
+
+def test_format_raw_stack_short():
+    stack = """
+        Traceback (most recent call last):
+          File "app.py", line 10, in <module>
+            my_func()
+          File "app.py", line 7, in my_func
+            raise ValueError("An error")
+        ValueError: An error
+        """
+    expected_result = textwrap.dedent(stack).strip()
+    result = format_raw_stack(stack)
+
+    assert result == expected_result
+
+
+@pytest.fixture
+def saved_function_key(mocker):
+    function_key = "saved_key"
+    mocker.patch("azure_functions.function_app.FUNCTION_KEY", function_key)
```

Comment:
This is a nit, but we typically organize all the `fixtures` up at the top of the file, and then the tests underneath. It is a little easier to scan top-to-bottom or skip past all the fixtures to get into the tests.